src/Conversion/ONNXToKrnl/Tensor/Gather.cpp

```diff
@@ -122,6 +127,32 @@ struct ONNXGatherOpLowering : public OpConversionPattern<ONNXGatherOp> {
           if (indicesMayBeNegative)
             index = index.selectOrSelf(index < zeroIE, index + axisDim);
 
+          // The Gather op is data dependent: the value of index should be
+          // within the input data size.
+          // Add runtime check if enableSafeCodeGen is set true
+          // Implementation comments vs. createGenerateRuntimeVerificationPass
+          // This check is according to onnx op semantics, not general bound
+          // check for memref. Implementation of RuntimeVerification could be
+          // borrowed. Slightly difference is that onnx semenatics check is for
+          // each dimension independently, not the final address is within
+          // the memref bound.
+          if (enableSafeCodeGen) {
+            // From onnx document:
+            // All index values are expected to be within bounds [-s, s-1]
+            // along axis of size s. It is an error if any of the index values
+            // are out of bounds.
+            // After the negative correction, the range should be [0, s-1]
+            Value upperBound = create.mem.dim(data, axisLit);
+            Value compareUpperBound =
+                create.math.slt(index.getValue(), upperBound);
+            rewriter.create<cf::AssertOp>(loc, compareUpperBound,
+                "indices of GatherOp is larger than the upper bound");
+            Value compareLowerBound =
+                create.math.sge(index.getValue(), zeroIE.getValue());
+            rewriter.create<cf::AssertOp>(loc, compareLowerBound,
```

Comment:
@chentong319 I know it is not optimized for speed, "anding" both condition and calling assert only once would speed the check up a bit. `create.math.andi(compareUpperBound, compareLowerBound)`.

Reply:
Separate the check to provide more accurate error message.